src/software/ai/hl/stp/tactic/receiver/receiver_fsm.h

```diff
@@ -0,0 +1,359 @@
+        // apply the offset in
+        if (lateral_vel.dot(shot_vector.rotate(Angle::quarter())) > 0)
+        {
+            // need to go clockwise
+            shot_offset = -shot_offset;
+        }
+        return shot_dir + shot_offset;
+    }
+
+    /**
+     * Figures out the location of the one-touch shot and orientation the robot should
+     * face
+     *
+     * @param robot The robot performing the one-touch
+     * @param ball The ball on the field
+     * @param best_shot_target The point to shoot at
+     */
+    static Shot getOneTouchShotPositionAndOrientation(const Robot& robot,
+                                                      const Ball& ball,
+                                                      const Point& best_shot_target)
+    {
+        double dist_to_ball_in_dribbler =
+            DIST_TO_FRONT_OF_ROBOT_METERS + BALL_MAX_RADIUS_METERS;
+        Point ball_contact_point =
+            robot.position() + Vector::createFromAngle(robot.orientation())
+                                   .normalize(dist_to_ball_in_dribbler);
+
+        // Find the closest point to the ball contact point on the ball's trajectory
+        Point closest_ball_pos = ball.position();
+        if (ball.velocity().length() >= BALL_MIN_MOVEMENT_SPEED)
+        {
+            closest_ball_pos =
+                closestPoint(ball_contact_point,
+                             Line(ball.position(), ball.position() + ball.velocity()));
+        }
+        Ray shot(closest_ball_pos, best_shot_target - closest_ball_pos);
+
+        Angle ideal_orientation      = getOneTouchShotDirection(shot, ball);
+        Vector ideal_orientation_vec = Vector::createFromAngle(ideal_orientation);
+
+        // The best position is determined such that the robot stays in the ideal
+        // orientation, but moves the shortest distance possible to put its contact
+        // point in the ball's path.
+        Point ideal_position =
+            closest_ball_pos - ideal_orientation_vec.normalize(dist_to_ball_in_dribbler);
+
+        return Shot(ideal_position, ideal_orientation);
+    }
+
+    /*
+     * Finds a shot that is greater than MIN_SHOT_NET_PERCENT_OPEN and
+     * respects MAX_DEFLECTION_FOR_ONE_TOUCH_SHOT for the highest chance
+     * of scoring with a one-touch shot. If neither of those are true, return a nullopt
+     *
+     * @param world The world to find a feasible shot on
+     * @param assigned_robot The robot that will be performing the one-touch
+     */
+    static std::optional<Shot> findFeasibleShot(const World& world,
+                                                const Robot& assigned_robot)
+    {
+        // Check if we can shoot on the enemy goal from the receiver position
+        std::optional<Shot> best_shot_opt = calcBestShotOnGoal(
+            world.field(), world.friendlyTeam(), world.enemyTeam(),
+            assigned_robot.position(), TeamType::ENEMY, {assigned_robot});
```

Comment:
why do we check for a best shot from the current position of the robot, and not its position when receiving the ball?

Reply:
the receiver is constantly adjusting,  so this is called in `update_onetouch` on every tick, so it gets it sort of right initially and then adjusts after the pass starts